[Issue 276] QEMU Agent configuration settings

mpywell opened the pull request:
Extend QEMU agent handling by introducing the `qemu_quest_agent` block

`qemu_guest_agent` adds the ability to set each of the QEMU Guest Agent's available settings:
- Enable/Disable the agent (replaces the `qemu_agent` bool)
- Setting the agent type
- Toggling freeze/thaw filesystems on backup
- Toggling fstrim after disk move or VM migration

Each added setting defaults to Proxmox's backend defaults when not configured.


Closes #276

mpywell committed:
common: add qemu agent settings block
- Extend QEMU agent handling by introducing qemu_quest_agent block, deprecating qemu_agent bool
- Enable setting the agent type
- Enable toggling freeze/thaw filesystems on backup
- Enable toggling fstrim after disk move or VM migration
- Each added setting defaults to Proxmox's backend defaults when not configured.

mpywell committed:
fix agent config test

lbajolet-hashicorp commented on builder/proxmox/common/config.go:
```diff
@@ -606,7 +648,27 @@ func (c *Config) Prepare(upper interface{}, raws ...interface{}) ([]string, []st
 	// Default qemu_agent to true
 	if c.Agent != config.TriFalse {
-		c.Agent = config.TriTrue
+		warnings = append(warnings, "qemu_agent is deprecated and will be removed in a future release. define QEMU agent settings in a qemu_guest_agent block instead")
```
I would suggest we only print that warning if `qemu_agent` is unset here, not if it's not explicitly `false`, otherwise we'll see it all the time unless we add `qemu_agent = false` to the template.

mpywell replied:
It occurred to me that we don't need to set a default value for `c.Agent` given the replacement `c.GuestAgent.Enabled` has a default set further down, so I've changed the logic to warn and convert the value only if the `c.Agent` is not Unset.

lbajolet-hashicorp commented on builder/proxmox/iso/config_test.go:
```diff
@@ -103,8 +103,8 @@ func TestBasicExampleFromDocsIsValid(t *testing.T) {
 	}
-	if b.config.Agent.True() != true {
-		t.Errorf("Expected Agent to be true, got %t", b.config.Agent.True())
+	if b.config.GuestAgent.Enabled.True() != true {
```
I would suggest simplifying this one; 
```suggestion
	if !b.config.GuestAgent.Enabled.True() {
```

mpywell replied:
Done, added to reroll

lbajolet-hashicorp reviewed (dismissed):
Thanks for this one @mpywell!

Overall LGTM, I left a couple of suggestions, feel free to address them and I'll come back for another round of review after that.

mpywell committed:
common: qemu_agent, qemu_guest_agent updates
- fix qemu_agent deprecation warnings and conversion
- invert freeze config option for qemu_guest_agent
- simplify test logic for TestBasicExampleFromDocsIsValid

mpywell commented on builder/proxmox/common/config.go:
```diff
@@ -207,6 +210,45 @@ type Config struct {
+	// Sets the Agent Type. Must be `isa` or `virtio`. Defaults to `virtio`
+	Type string `mapstructure:"type"`
+	// Enable freeze/thaw of guest filesystem on backup. Defaults to `true`
+	Freeze config.Trilean `mapstructure:"freeze"`
```
Makes sense, have rerolled with `disable_freeze` as a bool

mpywell commented:
Hi @lbajolet-hashicorp have rerolled this one and added comments in the review

DerLinkman commented:
This would be really nice to have merged :)

dreibh commented:
Indeed, this pull request would be useful. Currently, for building OpenBSD or NetBSD VMs, it requires to manually set the agent type to ISA.
